File: Python/343.Integer Break.py
#把一个2到58的正整数拆成几个数的和，使得他们的乘积最大，返回最大乘积。

#在n >3的情况下，我们处理一个数拆分成2，要么拆分成3，（4的话相当于2个2 ， 拆成1的话乘积太小了）
#可以说，拆成3的比拆成2的乘积大。 比如6的时候 2*2*2 < 3*3
#我们希望能尽可能的拆成3，然后才是2.
#所以，如果
#n % 3 == 0:  那么全部拆成3
#n % 3 == 1:  2个2剩下的为3    4*3^(x-1) > 1*3^x
#n % 3 == 2:  1个2剩下的为3
class Solution(object):
    def integerBreak(self, n):
        if n <= 3 :return n-1
        mod = n % 3
        if mod == 0: return int(pow(3, n / 3))
        if mod == 1: return int(4 * pow(3, (n - 4) / 3))
        return int(2 * pow(3, n / 3))

s = Solution()
print(s.integerBreak(15))

File: C++/Math/343.Integer Break.cpp
/*
给出整数n，把n分为至少两个数的和，让这几个数乘积最大。
返回最大乘积。n大于等于2，小于等于58。
*/
/*
思路：5以上的数都符合一个规律：要让乘积最大，需要尽量多的3。
那么，当n对3取余，为0时，显然这些3相乘，为1时，少一个3，多有一个4，
为2时，多乘个2。
数学推导：把n分为x份，乘积为x的n/x次方，求导，导数为0时的点为e，
靠近e的点只能是2和3。大的数都是3，小的数如4，分成2*2乘积最大。
*/
#include <iostream>
#include <cmath>
using namespace std;

class Solution {
public:
	int integerBreak(int n) {   
        if(n<4)    return n-1;
        if(n==4)    return 4;
        int m = n%3, r = n/3;
        if(m==0)    return pow(3,r);
        if(m==1)    return pow(3,r-1)*4;
        return pow(3,r)*2;
    }
};
int main()
{
	Solution s;
	cout << s.integerBreak(50) << endl;
}

